### PR TITLE
CATROID-258 Change names in formula fields if variable name changes

### PR DESCRIPTION
**Pull request by SchmidtChristian**
- add function to update all references to this variable in current project
- add testcase to test all formulabricks

### CODE REVIEW

**Comment by 84n4n4** [comment]
please remove binary files

File: catroid/src/main/java/org/catrobat/catroid/ProjectManager.java [hunk 1]
```diff
@@ -383,6 +384,41 @@ public static void removePermissionsFile(Project project) {
 		}
 	}
 
+	public static void updateUserDataReferences(String oldName, String newName,
+			UserData item) {
+		if (item instanceof UserList || item instanceof UserVariable) {
+			for (Scene scene : getInstance().getCurrentProject().getSceneList()) {
+				for (Sprite sprite : scene.getSpriteList()) {
+					for (Script script : sprite.getScriptList()) {
+						Brick scriptBrick = script.getScriptBrick();
```
Comment (84n4n4) [inline catroid/src/main/java/org/catrobat/catroid/ProjectManager.java]: List<Brick> flatList = new ArrayList<Brick>();
script.addToFlatList(flatList);

for (Brick brick : flatList) {
    if (brick instanceof FormulaBrick) { .....

File: catroid/src/test/java/org/catrobat/catroid/test/content/bricks/UpdateVariableInFormulaBrickTest.java [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+	private static final String VARIABLE_NAME_USERVARIABLE_FORMAT = "\"Test\" ";
+	private static final String VARIABLE_NAME_USERLIST_FORMAT = "*Test* ";
+	private static final String INVALID_NAME = "Abcd";
+	private static final String NEW_VARIABLE_NAME = "NewName";
+	private static final String NEW_VARIABLE_USERVARIABLE_FORMAT = "\"NewName\" ";
+	private static final String NEW_VARIABLE_USERLIST_FORMAT = "*NewName* ";
+
+	@Parameterized.Parameters(name = "{0}")
+	public static Iterable<Object[]> data() {
+		List<Object[]> parameters = new ArrayList<>();
```
Comment (84n4n4) [inline catroid/src/test/java/org/catrobat/catroid/test/content/bricks/UpdateVariableInFormulaBrickTest.java]: please exclude script bricks that are also formula bricks from this and have them tested in a seperate test class

File: catroid/src/test/java/org/catrobat/catroid/test/content/bricks/UpdateVariableInFormulaBrickTest.java [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+			formulaBrick.setFormulaWithBrickField(k, newFormula);
+		});
+
+		ProjectManager.updateUserDataReferences(INVALID_NAME, NEW_VARIABLE_NAME, userList);
+
+		map.forEach((k, v) -> {
+			assertEquals(v.getTrimmedFormulaString(CatroidApplication.getAppContext()),
+					VARIABLE_NAME_USERLIST_FORMAT);
+		});
+	}
```
Comment (84n4n4) [inline catroid/src/test/java/org/catrobat/catroid/test/content/bricks/UpdateVariableInFormulaBrickTest.java]: please test that treestructure is obeyed, by using CompositeBrickTest and CompositeBrickWithSecondaryListTest as template to test all composite bricks but in this case one Formula brick, eg. Set X Brick is sufficient.

**Comment by wslany** [comment]
Thanks!

**Comment by wslany** [comment]
@SchmidtChristian thanks for fixing the crash reported in [my comment above](https://github.com/Catrobat/Catroid/pull/3345#pullrequestreview-289105280) via https://github.com/Catrobat/Catroid/pull/3357

**Pull request merged**